feat: add protocol support to `stats/base/range`

jalajk3004 opened the pull request:
Resolves #5679 

## Description

> What is the purpose of this pull request?

This pull request:

-  Refactor and add protocol support

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   Resolves #5679 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

jalajk3004 committed:
refactor benchmark

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

jalajk3004 committed:
refactor

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
---

jalajk3004 commented:
@kgryte I have submitted a pull request addressing issue #5679, and I would love to contribute further to resolving this issue completely. I would appreciate your guidance on any necessary improvements or changes required to get this PR merged.

Please let me know how I can assist in the process. Looking forward to your feedback.

jalajk3004 committed:
approved

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: passed
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

jalajk3004 committed:
lint error

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/stats/base/range/README.md:
````diff
@@ -137,16 +129,14 @@ var v = range.ndarray( N, x, 2, 1 );
 ```javascript
 var randu = require( '@stdlib/random/base/randu' );
 var round = require( '@stdlib/math/base/special/round' );
-var Float64Array = require( '@stdlib/array/float64' );
+var linspace = require( '@stdlib/array/base/linspace' );
````
Use `@stdlib/random/array/discrete-uniform`.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/README.md:
```diff
@@ -137,16 +129,14 @@ var v = range.ndarray( N, x, 2, 1 );
-for ( i = 0; i < x.length; i++ ) {
-    x[ i ] = round( (randu()*100.0) - 50.0 );
-}
+x = linspace(-50, 50, 10);
+x = x.map(round);
```
This will then become

```suggestion
x = discreteUniform( 10, -50, 50, {
    'dtype': 'float64'
});
```

kgryte commented on lib/node_modules/@stdlib/stats/base/range/benchmark/benchmark.js:
```diff
@@ -21,8 +21,8 @@
 var bench = require( '@stdlib/bench' );
-var randu = require( '@stdlib/random/base/randu' );
 var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var linspace = require( '@stdlib/array/base/linspace' );
```
Use `@stdlib/random/array/uniform`, rather than `linspace`.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/benchmark/benchmark.ndarray.js:
```diff
@@ -22,6 +22,7 @@
 
 var bench = require( '@stdlib/bench' );
 var randu = require( '@stdlib/random/base/randu' );
+var linspace = require( '@stdlib/array/linspace' );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/docs/repl.txt:
```diff
@@ -1,92 +1,124 @@
+{{alias}}( N, x, strideX )
```
```suggestion

{{alias}}( N, x, strideX )
```

kgryte commented on lib/node_modules/@stdlib/stats/base/range/docs/repl.txt:
```diff
@@ -1,92 +1,124 @@
+    The `N` and stride parameters determine which elements in the strided 
+    array are accessed at runtime.
+
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/stats/base/range/docs/repl.txt:
```diff
@@ -1,92 +1,124 @@
 
-    Indexing is relative to the first index. To introduce an offset, use a typed
-    array view.
+    The `N` and stride parameters determine which elements in the strided 
```
This doesn't look properly wrapped at 80 chars.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/docs/repl.txt:
```diff
@@ -1,92 +1,124 @@
+
     Parameters
     ----------
+
```
Why are you adding all these blank lines? Please remove and follow our REPL text style guide.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/docs/repl.txt:
```diff
@@ -1,92 +1,124 @@
 
     Returns
     -------
+
```
Why?

kgryte commented on lib/node_modules/@stdlib/stats/base/range/docs/repl.txt:
```diff
@@ -1,92 +1,124 @@
     4.0
 
-    // Using `N` and `stride` parameters:
+
```
You need to go through and remove all these stray lines. Please study other repl.txt files.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/examples/index.js:
```diff
@@ -18,18 +18,10 @@
-for ( i = 0; i < x.length; i++ ) {
-	x[ i ] = round( (randu()*100.0) - 50.0 );
-}
+var x = linspace(-50, 50, 10);
```
This should be updated to match any future changes to the README.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/lib/accessors.js:
```diff
@@ -0,0 +1,82 @@
+		return NaN;
+	}
+	if ( N === 1 || stride === 0 ) {
+		const v = get( x, 0 );
```
We use ES5. All of this ES6+ usage needs to be changed. Please study the conventions in other files.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/lib/ndarray.js:
```diff
@@ -40,10 +41,13 @@ var isnan = require( '@stdlib/math/base/assert/is-nan' );
 */
-function range( N, x, stride, offset ) {
+function range( N, x, strideX, offsetX ) {
+	var obj = arraylike2object( x );
```
This is not what we want. Please study other implementations. Accessing directly in `data` is precisely what we want to avoid if `x` is an accessor array. That is the entire point of `accessors.js`.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/lib/range.js:
```diff
@@ -21,7 +21,8 @@
 // MODULES //
 
 var isnan = require( '@stdlib/math/base/assert/is-nan' );
```
This is no longer used.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/test/test.ndarray.js:
```diff
@@ -67,6 +68,57 @@ tape( 'the function calculates the range of a strided array', function test( t )
+
+	x = [ -3.0, -2.0, -1.0, 1.0, 2.0, 3.0 ];
+	v = range( x.length, toAccessorArray( x ), 1, 0 );
+	t.ok(isnan(v), 'returns expected value');
```
All these tests are obviously wrong. We shouldn't be returning `NaN`.

kgryte commented on lib/node_modules/@stdlib/stats/base/range/test/test.ndarray.js:
```diff
@@ -115,15 +166,56 @@ tape( 'the function supports a `stride` parameter', function test( t ) {
+
+	v = range( 4, toAccessorArray( x ), 2, 0 );
+
+	t.ok(isnan(v), 'returns expected value');
```
This is wrong.

kgryte reviewed (changes requested):
Left an initial round of comments. The implementation is currently wrong and the tests obviously so. I suggest doing a close study of other packages which have accessor protocol support to better understand what is desired.

jalajk3004 committed:
changes

jalajk3004 committed:
check

jalajk3004 committed:
check

jalajk3004 committed:
linting error

jalajk3004 committed:
repl

jalajk3004 committed:
done

jalajk3004 committed:
edit

jalajk3004 committed:
check

jalajk3004 committed:
repl

jalajk3004 committed:
check

jalajk3004 committed:
lint

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/range/test/test.ndarray.js:
```diff
@@ -1,106 +1,114 @@
-* See the License for the specific language governing permissions and
-* limitations under the License.
-*/
+ * @license Apache-2.0
```
@jalajk3004 Hey the license should not be tempered with here. Hence the CI error.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/range/test/test.ndarray.js:
```diff
@@ -1,106 +1,114 @@
-* See the License for the specific language governing permissions and
-* limitations under the License.
-*/
+ * @license Apache-2.0
```
```suggestion
* @license Apache-2.0
```
@jalajk3004 Hey the license should not be tempered with here. Hence the CI error.

jalajk3004 replied:
Thanks @Neerajpathak07, Error is solved.
But I am still facing the problem with Repl
![image](https://github.com/user-attachments/assets/6d37f789-e3ab-4808-9d22-aa89d7bf00d6)

Neerajpathak07 replied:
well that's because the expected and the returned values do not match. Check your implementation once again and do go through the `repl.txt` file to see if everything is in-line.

Neerajpathak07 commented on lib/node_modules/@stdlib/stats/base/range/docs/repl.txt:
```diff
@@ -30,25 +30,28 @@
-    > var N = {{alias:@stdlib/math/base/special/floor}}( x.length / 2 );
+    > var N = {{alias:@stdlib/math/base/special/floor}}( 3 / 2 );
     > var stride = 2;
     > {{alias}}( N, x, stride )
```
```suggestion
    > {{alias}}( 3, x, 2 )
```
You'll have to update the alias for the new values of `N` and `stride`. Same for L48 and L87
The CI error is becuz the expected and returned value do not match.

jalajk3004 committed:
license

jalajk3004 committed:
check

jalajk3004 committed:
repl

jalajk3004 committed:
-

jalajk3004 committed:
done

kgryte committed:
Merge branch 'develop' into range

Signed-off-by: Athan <kgryte@gmail.com>

gururaj1512 commented:
/stdlib help

gururaj1512 commented:
/stdlib merge

stdlib-bot committed:
Merge remote-tracking branch 'upstream/develop' into range

gururaj1512 committed:
docs: clean-up examples

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: clean repl

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: update implementation

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
bench: update benchmarks

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
fix: test cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
docs: minor clean-ups

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

gururaj1512 committed:
Update accessors.js

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

gururaj1512 committed:
Update ndarray.js

Signed-off-by: Gururaj Gurram <143020143+gururaj1512@users.noreply.github.com>

kgryte reviewed (approved):
LGTM
